008-auth.md

```diff
@@ -0,0 +1,254 @@
+of events, etc.
+
+Include any diagrams as appropriate, favoring text-based sources others can
+easily edit like mermaid, UML, etc. where possible, with external files sharing
+the same filename prefix as this document.
+
+# Examples
+
+Detail one or more hypothetical but plausible real-world scenarios where this
+RFC would offer a tangible benefit.
+
+# Drawbacks
+
+Provide a few bullets honestly detailing the downsides of implementing this.
+
+Will it slow down development? Cost more resources? Etc.
+
+# Alternatives
+
+List any alternatives others have tried, including the current solution if
+there is one, to help others quickly catch up on how we got here.
+
+# Scope
+
+List a few items that are clearly in scope, anything that we want to do that is
+explicitly deferred to future work.
+
+## Questions
+
+What open questions still exist that we can't answer right now?
+
+
+
+
+
+
+# Client authentication
+
+
+## Phase 1
+
+First, we need to break apart the `talosconfig` from the authentication PKI.
+This allows us to, among other things: 
+ - encourage the distribution of `talosconfig` without secret material included
+ - encourage the disuse of root admin keys for normal operation
+ - encourage the use of user-specific secret material which is never sent over
+   the wire
+ - use a common user keypair across multiple Talos clusters
+
+Some minor additions to `talosconfig` will be useful:
+ - top-level `defaultAuthIdentity` reference
+ - cluster-level `authIdentity` reference
+ - cluster-level `clusterId` reference
+ - cluster-level `certs` list
+
+Perhaps add a `-i` option to `talosctl` which tells Talos to use the
+given auth identity.
+
+Construct and use `~/.config/talos/auth` (or ~/.talos/auth, for non-XDG-standard
+machines) to house the Talos identities on the client machine.
+
+Modify `talosctl` to use the search the auth directory to resolve authentication
+material when connecting to Talos.
+
+Proposed structure of `~/.config/talos/auth`:
+ - `id_<name>.key` : private key of the given id
+ - `id_<name>.pub` : public key of the given id
+
+Proposed CLI subcommand:  `auth`:
+ - `talosctl auth approve` : approve a user's certificate request
+ - `talosctl auth request` : create a request to access a cluster
+ - `talosctl auth gen[erate] id` : generate a user keypair
+
+ - `talosctl config strip-auth` : strip the secret auth material from `talosconfig`
+ - `talosctl config import-cert` : import a signed certificate
+
+
+## Phase 2
+
+Focusing on the most basic operational requirements, we need to let `talosctl`:
+ - Generate a client keypair
+ - Generate a CSR from a client keypair
+ - Generate a signed certificate from a CSR
+
+In order to make things work a little better for user-oriented
+
+## Addendum
+
+`talosctl kubeconfig` should offer RBAC-restricted versions:  namespace
+restrictions, for instance.
+Require `--admin` to generate an admin cert
+Maybe enforce one of `--namespaces` or `--admin`
```

Comment:
Why? kube-apiserver can already be configured with OIDC, and assuming you install ClusterRole and the other necessary RBAC in the cluster, there's no need to hand out credentials at all - the kubeconfig would be hand-crafted and point to whatever auth-helper, that'll pick up creds from ambient credentials, or ask the user for it.

Reply:
Fair, but this change does not mandate OIDC.  It is more generalised than that.  Further, it does not impose any restrictions or bindings of authentication between Talos and Kubernetes.

Reply:
That's true, but part of the motivation here is to encourage people to use bounded credentials; to make tooling to do the "right thing" easier.
Your point is reasonable, though:  people _shouldn't_ really be using `talosctl kubeconfig` to extract Kubernetes credentials as a matter of standard course.
However, some of our tooling internally _does_ rely on exactly that.